packages/transaction-controller/src/utils/history.ts

```diff
@@ -8,39 +8,47 @@ import type {
 } from '../types';
 
 /**
- * Add initial history snapshot to the provided transactionMeta history.
+ * Build a new version of the provided transaction with an initial history
+ * entry, which is just a snapshot of the transaction.
  *
  * @param transactionMeta - TransactionMeta to add initial history snapshot to.
+ * @returns A copy of `transactionMeta` with a new `history` property.
  */
-export function addInitialHistorySnapshot(transactionMeta: TransactionMeta) {
+export function addInitialHistorySnapshot(
+  transactionMeta: TransactionMeta,
+): TransactionMeta {
   const snapshot = snapshotFromTransactionMeta(transactionMeta);
-  transactionMeta.history = [snapshot];
+  return { ...transactionMeta, history: [snapshot] };
 }
 
 /**
- * Compares and adds history entry to the provided transactionMeta history.
+ * Builds a new version of the transaction with a new history entry if
+ * it has a `history` property, or just returns the transaction.
  *
  * @param transactionMeta - TransactionMeta to add history entry to.
  * @param note - Note to add to history entry.
+ * @returns A copy of `transactionMeta` with a new `history` entry if it has a
+ * `history`, or `transactionMeta` unchanged.
```

Comment:
```suggestion
 * @returns A copy of `transactionMeta` with a new `history` entry if it has an
 * existing non-empty `history` array
```

or something like that?

Reply:
Yes — updated here. [de1f551](https://github.com/MetaMask/core/pull/3827/commits/de1f551f6b810e1e2b95c07a9157d6314c014240)